### PR TITLE
[BEAM-2597] Flink batch ExecutableStage operator

### PR DESCRIPTION
**Pull request by bsidhom**
This is a first pass at an implementation of the `FlinkExecutableStageFunction`.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `./gradlew build` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by bsidhom** [comment]
I've left some TODOs in place as questions for reviewers.

Unfortunately, it turns out that we need a few more interfaces for bundle management due to serialization. Flink requires that `Function` instances be `Serializable` as this is how they are distributed within a cluster. In order to make this unit-testable, I had to introduce (stateless) serializable factories to produce the stateful, non-serializable instances that need to be managed within the TaskManagers themselves.

The exact shape of `FlinkBundleFactory` is pending [BEAM-3673](https://issues.apache.org/jira/browse/BEAM-3673)/[BEAM-3327](https://issues.apache.org/jira/browse/BEAM-3327). `CloseableDistributedCache` will hopefully go away with the right abstractions on top of ArtifactSource. I wasn't sure how we wanted that part to look as it's not the interesting part of stage execution.

Additionally, since the state handler does not quite handle coders correctly due to inconsistencies with ProcessbundleDescriptors/fusion, that part may be better as a part of a follow-up PR. I'm happy to make that a skeleton/stub as I did with `ExecutableStageFunction` originally.

R: @jkff @tgroh 
CC: @axelmagn @angoenka @lukecwik

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/OutputReceiverFactory.java [hunk 1]
```diff
@@ -18,14 +18,16 @@
...
+
 /**
  * A factory that can create output receivers during an executable stage.
  */
 public interface OutputReceiverFactory {
   /**
    * Get a new {@link RemoteOutputReceiver} for an output PCollection.
    *
-   * This call should only be invoked once per PCollectionId per factory.
+   * <p>This call should only be invoked once per PCollection id per factory.
```
Comment (tgroh) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/OutputReceiverFactory.java]: This restriction is only sort of true, as I would understand it. It's probably not strictly required, but some implementations may require this.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java [hunk 2]
```diff
@@ -38,36 +53,115 @@
...
+      }
+    }
+    // NOTE: RemoteBundle.close() blocks on completion of all data receivers. This is necessary to
+    // safely reference the partition-scoped Collector from receivers.
   }
 
   @Override
   public void close() throws Exception {
-    throw new UnsupportedOperationException();
+    // TODO: Wrap close calls to ensure all calls are attempted?
```
Comment (tgroh) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]:
```
try (AutoCloseable cache = distributedCache;
     AutoCloseable factory = stageBundleFactory;) {}
```

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/OutputReceiverFactory.java (hunk 1 above)
Comment (bsidhom) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/OutputReceiverFactory.java]: It's not necessary in all cases, but this is something we want for runner implementors.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java (hunk 2 above)
Comment (bsidhom) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]: Clever (if strange looking).

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBundleFactory.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  public static FlinkBundleFactory getInstance() {
+    return Holder.INSTANCE;
+  }
+
+  /** Gets or creates a {@link JobBundleFactory}. */
+  public JobBundleFactory getJobBundleFactory(JobInfo jobInfo, ArtifactSource artifactSource) {
+    throw new UnsupportedOperationException();
+  }
+
+  private static class Holder {
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBundleFactory.java]: Is this class necessary?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkBatchStateRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowedValue.WindowedValueCoder;
+import org.apache.beam.sdk.values.KV;
+import org.apache.flink.api.common.functions.RuntimeContext;
+
+/**
+ * {@link StateRequestHandler} that uses a Flink {@link RuntimeContext} to access Flink broadcast
+ * variable that represent side inputs.
+ */
+class FlinkBatchStateRequestHandler implements StateRequestHandler {
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkBatchStateRequestHandler.java]: (not reviewing this one since you said it's subject to redesign due to coder issues)

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java [hunk 5]
```diff
@@ -38,36 +53,117 @@
...
   public void open(Configuration parameters) throws Exception {
-    throw new UnsupportedOperationException();
+    ExecutableStage executableStage = ExecutableStage.fromPayload(stagePayload);
+    runtimeContext = getRuntimeContext();
+    // NOTE: It's safe to reuse the state handler between partitions because each partition uses the
+    // same backing runtime context and broadcast variables. We use checkState below to catch errors
+    // in backward-incompatible Flink changes.
+    stateRequestHandler = stateHandlerFactory.forStage(executableStage, runtimeContext);
+    DistributedCachePool cachePool = cachePoolFactory.forJob(jobInfo.jobId());
+    distributedCacheCloser = cachePool.addCacheToPool(runtimeContext.getDistributedCache());
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]: * Weren't we going to have a pool of ArtifactSource rather than a pool of DistributedCache?
* This pooling business is very funky (remember how much in-person discussion it took to remember why we even need it), could you add some comments about what's going on here?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java [hunk 6]
```diff
@@ -38,36 +53,117 @@
...
+  private transient StageBundleFactory stageBundleFactory;
+  private transient AutoCloseable distributedCacheCloser;
 
   public FlinkExecutableStageFunction(
       RunnerApi.ExecutableStagePayload stagePayload,
-      Struct pipelineOptions,
-      Map<String, Integer> outputMap) {
+      JobInfo jobInfo,
+      Map<String, Integer> outputMap,
+      SerializableSupplier<FlinkBundleFactory> bundleFactorySupplier,
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]: Supplier of a factory seems like too many layers of indirection, can the bundle factory itself be serializable?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/DistributedCachePool.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi.Manifest;
+import org.apache.beam.runners.fnexecution.artifact.ArtifactSource;
+import org.apache.flink.api.common.cache.DistributedCache;
+
+/**
+ * A pool of {@link DistributedCache DistributedCaches} to use as artifact sources. A cache must be
+ * registered before artifacts are requested from it.
+ */
+@ThreadSafe
+public class DistributedCachePool implements ArtifactSource {
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/DistributedCachePool.java]: Perhaps call this ArtifactSourcePool?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/DistributedCachePool.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+   */
+  public interface Factory extends Serializable {
+    /** Gets or creates a cache pool for the given job id. */
+    DistributedCachePool forJob(String jobId);
+  }
+
+  /** Retrieves the default distributed cache pool factory. */
+  public static Factory defaultFactory() {
+    return (jobId) -> {
+      throw new UnsupportedOperationException();
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/DistributedCachePool.java]: Do you plan to implement these in the current PR?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/DistributedCachePool.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    };
+  }
+
+  private DistributedCachePool() {}
+
+  /**
+   * Adds a new cache to the pool. When the returned {@link AutoCloseable} is closed, the given
+   * cache will be removed from the pool.
+   */
+  public AutoCloseable addCacheToPool(DistributedCache distributedCache) {
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/DistributedCachePool.java]: This could take an ArtifactSource

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java [hunk 10]
```diff
@@ -38,36 +53,117 @@
...
+    ExecutableStage executableStage = ExecutableStage.fromPayload(stagePayload);
+    runtimeContext = getRuntimeContext();
+    // NOTE: It's safe to reuse the state handler between partitions because each partition uses the
+    // same backing runtime context and broadcast variables. We use checkState below to catch errors
+    // in backward-incompatible Flink changes.
+    stateRequestHandler = stateHandlerFactory.forStage(executableStage, runtimeContext);
+    DistributedCachePool cachePool = cachePoolFactory.forJob(jobInfo.jobId());
+    distributedCacheCloser = cachePool.addCacheToPool(runtimeContext.getDistributedCache());
+    FlinkBundleFactory flinkBundleFactory = bundleFactorySupplier.get();
+    // TODO: Do we really want this layer of indirection when accessing the stage bundle factory?
```
Comment (jkff) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]: I think for now it's fine - it would look less awkward if Flink had hooks for stage startup/shutdown separate from job startup/shutdown.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBundleFactory.java (hunk 3 above)
Comment (bsidhom) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBundleFactory.java]: Removed.

Note: I originally thought this class might do more work and wanted lazy initialization. However, given that the bundle factory itself must have TaskManager-local state, this cannot be passed directly to the `FlinkExecutableStage` at construction. Instead, any heavy lifting will happen in the `SerializableSupplier` that produces this. `getInstance` will probably change to return a `SerializableSupplier<FlinkBundleFactory>`. This is effectively a FactoryFactory(FactoryFactoryFactory).

It's unfortunate we need this many layers of indirection for object creation, but it's necessary for unit testing and distributed execution.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java (hunk 6 above)
Comment (bsidhom) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]: I noticed this comment after leaving another one above (on FlinkBundleFactory).

tl;dr That would be nice in theory, but the FlinkBundleFactory should really only be instantiated on worker nodes and will have machine-local state. I think it would be wrong to make it serializable.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java (hunk 5 above)
Comment (bsidhom) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]: Done. I've added those comments in ArtifactSourcePool.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java (hunk 10 above)
Comment (bsidhom) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]: OK. I'll leave the TODO in place as this is as-yet unanswered.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/DistributedCachePool.java (hunk 7 above)
Comment (bsidhom) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/DistributedCachePool.java]: I've gone ahead and made that change here but have left out the actual DistributedCache wrapper stub. I did this to avoid introducing yet another factory requirement for `FlinkExecutableStageFunction`. We might want to talk offline about the structure of this and what we can do to make things cleaner.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkBatchStateRequestHandler.java (hunk 4 above)
Comment (bsidhom) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkBatchStateRequestHandler.java]: I've gone ahead and removed the implementation but left the class stub.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java (hunk 6 above)
Comment (jkff) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java]: Hm I wonder if it makes sense then to encapsulate these three things into a "FlinkExecutableStageContext" and have a ContextFactory, so that at least we avoid the proliferation of factories.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunctionTest.java [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+    FlinkExecutableStageFunction<Integer> function = getFunction(Collections.emptyMap());
+    function.open(new Configuration());
+    // Change runtime context.
+    function.setRuntimeContext(Mockito.mock(RuntimeContext.class));
+    thrown.expect(Matchers.instanceOf(IllegalStateException.class));
+    function.mapPartition(Collections.emptyList(), collector);
+  }
+
+  @Test
+  public void inputsAreSentInOrder() throws Exception {
```
Comment (jkff) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunctionTest.java]: Is this an important property to verify? PCollections are unordered, as are bundles.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunctionTest.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    FlinkExecutableStageFunction<Integer> function = getFunction(outputTagMap);
+    function.open(new Configuration());
+
+    CompletableFuture<OutputReceiverFactory> receiverFactoryFuture = new CompletableFuture<>();
+
+    @SuppressWarnings("unchecked")
+    RemoteBundle<Integer> bundle = Mockito.mock(RemoteBundle.class);
+    // Capture the real receiver factory when a bundle is requested.
+    when(stageBundleFactory.<Integer>getBundle(any(), any()))
+        .thenAnswer(
```
Comment (jkff) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunctionTest.java]: Not sure if mocks are the best way to handle this, maybe anonymous classes would be better because you would't be losing static typing.

Comment (bsidhom) [reply runners/flink/src/test/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunctionTest.java]: I agree. However, it's painful to write this full set of nested factories uniquely for each test case. I think this will be better once we only have a single Context that handles all the auxiliary details, but I wan to figure out exactly what that looks like.

**Comment by bsidhom** [comment]
I've removed the nested factories and wrapped all of the interesting utilities into `FlinkExecutableStageContext`. This context itself (along with actual environment management) still needs to be implemented.

PTAL

**Comment by bsidhom** [comment]
Run Java Precommit

**Comment by bsidhom** [comment]
I'm not sure why we're seeing a failure reported here: https://builds.apache.org/job/beam_PreCommit_Java_GradleBuild/5264/testReport/junit/org.apache.beam.runners.fnexecution.data/GrpcDataServiceTest/testMessageReceivedBySingleClientWhenThereAreMultipleClients/

`GrpcDataServiceTest` runs successfully on my machine. I'm having Jenkins rerun to see if this was an unexpected flake.

**Comment by jkff** [comment]
Yeah your PR doesn't change any code that could possibly affect this test - the entire PR is in runners/flink, except for a new class and a change in comments. The test also has passed in the current pending run according to the gradle output. I'm gonna just merge it.
